src/daemon/daemon.cpp

```diff
@@ -2070,40 +2072,19 @@ QFutureWatcher<mp::Daemon::AsyncOperationStatus>* mp::Daemon::create_future_watc
     return future_watcher;
 }
 
-grpc::Status mp::Daemon::async_wait_for_ssh_for(const VirtualMachine::UPtr& vm)
-{
-    try
-    {
-        vm->wait_until_ssh_up(up_timeout);
-    }
-    catch (const std::exception& e)
-    {
-        return grpc::Status(grpc::StatusCode::FAILED_PRECONDITION, e.what(), "");
-    }
-
-    return grpc::Status::OK;
-}
-
 template <typename Reply>
-mp::Daemon::AsyncOperationStatus
-mp::Daemon::async_wait_for_ssh_and_start_mounts(grpc::ServerWriter<Reply>* server, const std::vector<std::string>& vms,
-                                                std::promise<grpc::Status>* status_promise)
+std::string mp::Daemon::async_wait_for_ssh_and_start_mounts_for(const std::string& name,
+                                                                grpc::ServerWriter<Reply>* server)
```

Comment:
Using return of an empty string to indicate success, and non-empty to indicate error, bothers me greatly.

I know in standard C++, the recommended paradigm for passing exceptions threads is to use `exception_ptr` to  make a copy of the exception and pass it to the future and rethrow it. Something like this:
https://stackoverflow.com/questions/233127/how-can-i-propagate-exceptions-between-threads

Could you give it a try?

Reply:
@gerboland,

Sure, I understand what you are saying, but in this case, it's really not a case of exception detection per se, but more about passing **all** error strings back to the client via `grpc_status_for()`.  You can see many examples of where we do that in the daemon code.

In `async_wait_for_ready_all()`, `start_synchronizer` is responsible for coalescing the error strings returned back from the thread(s) and then call `grpc_status_for()` on the full error string and based on that, the proper grpc status is set along with the error string.

Does that make sense or are you still bothered by it?  I kind of feel we will accomplish the same exact thing by trying to pass the exception via the thread as to what we do here at the cost of complication.

Reply:
I'm still bothered by it, but could live with it if we make it more explicit that the string returned by this method are error string(s)? Because right now it is not obvious what this is returning.

Even a `typedef std::string errorstring` would help satisfy me.

Reply:
Ok, I made at least the change to define the return type to be something more meaningful.  Perhaps later we can explore exception passing in the future if we find there is a need for it.